File: wtf_ruby/hangman.rb
require 'net/https'
require 'uri'
require 'json'
require 'logger'
require 'set'

class Hangman
  def initialize (option)
    @logger = Logger.new(STDOUT)
    @option = option
    @current_number_of_guess = 0
    @dict = []
    @current_dict_level = 0
    @current_sub_dict = []
    @guessed_char = []
  end

  def play
    load_all_dict
    start_game
  end

  private

  def next_round
    @guessed_char = []
    @current_dict_level = 0
    give_me_a_word
    build_sub_dict_of_len(@current_word.length)
    next_guess
  end

  def next_guess
    if @pass_current
      @pass_current = false
      next_round
    else
      guess = make_a_guess
      if guess_finished?
        if all_finished?
          get_result
        else
          next_round
        end
      elsif failed?
        get_result
        next_round
      elsif guess_wrong?
        @guessed_char << guess
        reduce_cur_sub_dict_by_chars([guess])
        next_guess
      else
        reduce_cur_sub_dict_by_word(@current_word)
        next_guess
      end
    end
  end

  def start_game
    res_obj = send_request(get_start_game_req(@option[:playerId]))
    @logger.info("Start game response: #{res_obj}")
    @session_id = res_obj['sessionId']
    @number_of_words_to_guess = res_obj['data']['numberOfWordsToGuess']
    @number_of_guess_allowed_for_each_word = res_obj['data']['numberOfGuessAllowedForEachWord']
    next_round
  end

  def give_me_a_word
    @current_number_of_guess = 0
    res_obj = send_request(get_give_me_a_word_req(@session_id))
    @logger.info("Give me a word response: #{res_obj}")
    @current_word = res_obj['data']['word']
    @total_word_count = res_obj['data']['totalWordCount']
    @wrong_guess_count_of_current_word = res_obj['data']['wrongGuessCountOfCurrentWord']
  end

  def make_a_guess
    next_guess = guess
    @logger.info("Next guess: #{next_guess}")
    @current_number_of_guess += 1
    res_obj = send_request(get_make_a_guess_req(@session_id, next_guess))
    @logger.info("Make a guess response: #{res_obj}")
    @previous_word = @current_word
    @current_word = res_obj['data']['word']
    @total_word_count = res_obj['data']['totalWordCount']
    @wrong_guess_count_of_current_word = res_obj['data']['wrongGuessCountOfCurrentWord']
    next_guess
  end

  def get_result
    res_obj = send_request(get_get_result_req(@session_id))
    @logger.info("Get result response: #{res_obj}")
    @correct_word_count = res_obj['data']['correctWordCount']
    @total_word_count = res_obj['data']['totalWordCount']
    @total_wrong_guess_count = res_obj['data']['totalWrongGuessCount']
    @score = res_obj['data']['score']
  end

  def submit_result
    send_request(get_submit_result_req(@session_id))
  end

  def send_request (request_body)
    uri = URI(@option[:host])
    req = Net::HTTP::Post.new(uri)
    req['Content-Type'] = 'application/json; charset=utf-8'
    req['Content-Length'] = request_body.length

    res = Net::HTTP.start(uri.hostname, uri.port, :use_ssl => uri.scheme == 'https') do |protocol|
      protocol.request(req, request_body.to_json)
    end

    case res
      when Net::HTTPSuccess
        return JSON.parse(res.body)
      else
        fail res.value
    end
    #timeout todo

  end

  def get_give_me_a_word_req(session_id)
    {:sessionId => session_id, :action => 'nextWord'}
  end

  def get_make_a_guess_req(session_id, guess)
    {:sessionId => session_id, :action => 'guessWord', :guess => guess}
  end

  def get_start_game_req(player_id)
    {:playerId => player_id, :action => 'start_game'}
  end

  def get_get_result_req(session_id)
    {:sessionId => session_id, :action => 'get_result'}
  end

  def get_submit_result_req(session_id)
    {:sessionId => session_id, :action => 'submit_result'}
  end

  def load_all_dict
    @option[:dictList].each {
        |name| load_dict(name)
    }
  end

  def load_dict(name)
    dict_index = @dict.length
    @dict[dict_index] = []
    File.open(File.join(@option[:dictPath], name)) do |file|
      file.each do |word|
        unless @dict[dict_index][word.length]
          @dict[dict_index][word.length] = []
        end
        @dict[dict_index][word.length] << word
      end
    end
    @logger.info("Dictionary loaded: #{name}")
  end

  def upgrade_sub_dict
    @current_dict_level += 1
    @logger.info("Dictionary upgraded to level: #{@current_dict_level}")
    build_sub_dict_of_len(@current_word.length)
    reduce_cur_sub_dict_by_word(@current_word)
    reduce_cur_sub_dict_by_chars(@guessed_char)
  end

  def can_upgrade_dict?
    @dict.length > @current_dict_level + 1
  end

  def build_sub_dict_of_len(length)
    unless @dict[@current_dict_level][length]
      @dict[@current_dict_level][length] = []
    end
    @current_sub_dict = @dict[@current_dict_level][length].clone
  end

  def reduce_cur_sub_dict_by_word(to_be_guessed)
    negated_set_reg = '[^'
    to_be_guessed.each_char do |char|
      negated_set_reg << char if (char != '*') && !negated_set_reg.include?(char)
    end
    negated_set_reg << ']'
    regex = to_be_guessed.gsub('*', negated_set_reg)
    @current_sub_dict.select! { |word| word.match(regex) }
    upgrade_dict_if_need
  end

  def reduce_cur_sub_dict_by_chars(characters)
    @current_sub_dict.select do |word|
      for i in 0..characters.length - 1
        next false if word.include?(characters[i])
      end
      next true
    end
    upgrade_dict_if_need
  end

  def upgrade_dict_if_need
    if @current_sub_dict.empty?
      @logger.warn("No candidate in dictionary: #{@current_word}")
      if can_upgrade_dict?
        upgrade_sub_dict
      else
        @pass_current = true
      end
    end
  end

  def get_most_possible_char (to_be_guessed)
    alpha_map = {'A' => 0}
    result = 'A'
    tmp = Set.new
    @current_sub_dict.each do |word|
      tmp.clear
      word.each_char do |char|
        tmp << char unless to_be_guessed.include?(char) || char == "\n"
      end
      tmp.each do |alpha|
        alpha_map[alpha] = alpha_map[alpha] ? alpha_map[alpha] + 1 : 1
        result = alpha if alpha_map[alpha] > alpha_map[result]
      end
    end
    result
  end

  def guess
    get_most_possible_char(@current_word)
  end

  def guess_finished?
    !@current_word.include?('*')
  end

  def guess_wrong?
    @current_word == @previous_word
  end

  def all_finished?
    @total_word_count == @number_of_words_to_guess
  end

  def failed?
    @wrong_guess_count_of_current_word == @number_of_guess_allowed_for_each_word
  end
end

File: hangman.js
var https = require('https');
var fs = require('fs');
var path = require('path');

module.exports = Hangman;

/**
 *
 * @param {Object} option
 * @constructor
 */
function Hangman(option) {
  if (!option) {
    throw new Error('option is needed');
  }
  this.option_ = option;
  this.currentNumberOfGuess_ = 0;
  this.gameContext_ = {};
  this.dict_ = [];
  this.currentDictLevel_ = 0;
  this.currentSubDict_ = [];
  this.guessedChar_ = [];
}

/**
 * auto play the game!
 */
Hangman.prototype.play = function () {
  var self = this;
  self.loadAllDict();
  self.startGame(function () {
    self.nextRound();
  });
};

/**
 * start next round
 */
Hangman.prototype.nextRound = function () {
  var self = this;
  self.guessedChar_ = [];
  self.currentDictLevel_ = 0;
  self.giveMeAWord(function () {
    self.buildSubDictOfLength(self.gameContext_['currentWord'].length);
    self.nextGuess();
  });
};

/**
 * give an another guess
 */
Hangman.prototype.nextGuess = function () {
  var self = this;
  if (self.passCurrent_) {
    self.passCurrent_ = false;
    self.nextRound();
  }
  else {
    self.makeAGuess(function (guess) {
      if (self.guessFinish()) {
        if (self.allFinished()) {
          self.getResult(function () {
            process.exit(0);
          });
        }
        else {
          self.nextRound();
        }
      } else if (self.failed()) {
        self.getResult();
        self.nextRound(); //////
      } else if (self.guessWrong()) {
        self.guessedChar_.push(guess);
        self.reduceCurrentSubDictByChars(guess);
        self.nextGuess();
      } else {
        self.reduceCurrentSubDictByWord(self.gameContext_['currentWord']);
        self.nextGuess();
      }
    })
  }
};

/**
 * start the game!
 * @param {function=} callback
 */
Hangman.prototype.startGame = function (callback) {
  var self = this;
  this.sendRequest(this.generateStartGameRequest(self.option_.playerId), function (err, response) {
    if (err) {
      throw err;
    }
    console.log('start game response: ' + response);
    var responseObj = JSON.parse(response);
    self.gameContext_['sessionId'] = responseObj['sessionId'];
    self.gameContext_['numberOfWordsToGuess'] = responseObj['data']['numberOfWordsToGuess'];
    self.gameContext_['numberOfGuessAllowedForEachWord'] = responseObj['data']['numberOfGuessAllowedForEachWord'];
    if (callback) {
      callback();
    }
  })
};

/**
 * give me a word!
 * @param {function=} callback
 */
Hangman.prototype.giveMeAWord = function (callback) {
  var self = this;
  this.currentNumberOfGuess_ = 0;
  this.sendRequest(this.generateGiveMeAWordRequest(this.gameContext_['sessionId']), function (err, response) {
    if (err) {
      throw err;
    }
    console.log('give me a word response:' + response);
    var responseObj = JSON.parse(response);
    self.gameContext_['currentWord'] = responseObj['data']['word'];
    self.gameContext_['totalWordCount'] = responseObj['data']['totalWordCount'];
    self.gameContext_['wrongGuessCountOfCurrentWord'] = responseObj['data']['wrongGuessCountOfCurrentWord'];
    if (callback) {
      callback();
    }
  })
};

/**
 * make a guess, callback will take the next guess as argument
 * @param {function=} callback
 */
Hangman.prototype.makeAGuess = function (callback) {
  var self = this;
  var nextGuess = this.guess();
  console.log('next guess: ' + nextGuess);
  this.currentNumberOfGuess_ = this.currentNumberOfGuess_ + 1;
  this.sendRequest(this.generateMakeAGuessRequest(this.gameContext_['sessionId'], nextGuess), function (err, response) {
    console.log('make a guess response:' + response);
    var responseObj = JSON.parse(response);
    self.gameContext_['previousWord'] = self.gameContext_['currentWord'];
    self.gameContext_['currentWord'] = responseObj['data']['word'];
    self.gameContext_['totalWordCount'] = responseObj['data']['totalWordCount'];
    self.gameContext_['wrongGuessCountOfCurrentWord'] = responseObj['data']['wrongGuessCountOfCurrentWord'];
    if (callback) {
      callback(nextGuess);
    }
  });
};

/**
 * get result
 * @param {function=} callback
 */
Hangman.prototype.getResult = function (callback) {
  var self = this;
  this.sendRequest(this.generateGetResultRequest(this.gameContext_['sessionId']), function (err, response) {
    if (err) {
      throw err;
    }
    console.log('get result response:' + response);
    var responseObj = JSON.parse(response);
    self.gameContext_['correctWordCount'] = responseObj['data']['correctWordCount'];
    self.gameContext_['totalWordCount'] = responseObj['data']['totalWordCount'];
    self.gameContext_['totalWrongGuessCount'] = responseObj['data']['totalWrongGuessCount'];
    self.gameContext_['score'] = responseObj['data']['score'];
    if (callback) {
      callback();
    }
  })
};

/**
 * submit result
 */
Hangman.prototype.submitResult = function () {
  this.sendRequest(this.generateSubmitResultRequest(this.gameContext_['sessionId']), function (err, response) {

  })
};

/**
 * send request
 * @param {Object} requestBody
 * @param {function=} callback
 */
Hangman.prototype.sendRequest = function (requestBody, callback) {
  var self = this;
  var requestStr = JSON.stringify(requestBody);
  var postOptions = {
    host: self.option_.host,
    path: self.option_.path,
    port: self.option_.port,
    method: 'POST',
    headers: {
      'Content-Type': 'application/json; charset=utf-8',
      'Content-Length': requestStr.length
    }
  };

  // Set up the request
  var req = https.request(postOptions, function (res) {
    res.setEncoding('utf8');
    var responseBody = '';
    res.on('data', function (chunk) {
      responseBody += chunk;
    });

    res.on('end', function () {
      if (res.statusCode >= 400) {
        console.warn('sth wrong in server, response status code: ' + res.statusCode);
        console.warn('response body: ' + responseBody);
        console.warn('request body:' + requestStr);

      } else if (callback) {
        //err the first parameter
        if (!responseBody) {
          console.warn('empty response');
          console.trace();
          self.sendRequest(requestBody, callback); //retry
        } else {
          callback(null, responseBody);
        }
      }
    });

    res.on('error', function (e) {
      console.error(e);
    })
  });

  req.on('error', function (e) {
    if (e.code === 'ECONNRESET') {
      console.warn('drop timeout request.');
    }
    //else if(e.code === 'ETIMEDOUT'){
    //  req.abort();
    //  console.warn('request timeout, try again: ' + JSON.stringify(requestBody));
    //  self.sendRequest(requestBody, callback); //retry
    //}
    else {
      console.error(e);
    }
  });

  req.setTimeout(10000, function () {
    req.abort();
    console.warn('request timeout, try again: ' + JSON.stringify(requestBody));
    self.sendRequest(requestBody, callback); //retry
  });

  // post the data
  req.write(requestStr);
  req.end();
};

/**
 * get give me a word request
 * @param {string} sessionId
 * @return {{sessionId: string, action: string}}
 */
Hangman.prototype.generateGiveMeAWordRequest = function (sessionId) {
  return {
    'sessionId': sessionId,
    'action': 'nextWord'
  }
};

/**
 * get make a guess request
 * @param sessionId
 * @param guess
 * @return {{sessionId: string, action: string, guess: string}}
 */
Hangman.prototype.generateMakeAGuessRequest = function (sessionId, guess) {
  return {
    'sessionId': sessionId,
    'action': 'guessWord',
    'guess': guess
  }
};

/**
 * get start game request
 * @param {string} playerId
 * @return {{playerId: string, action: string}}
 */
Hangman.prototype.generateStartGameRequest = function (playerId) {
  return {
    'playerId': playerId,
    'action': 'startGame'
  };
};

/**
 * get get result request
 * @param {string} sessionId
 * @return {{sessionId: string, action: string}}
 */
Hangman.prototype.generateGetResultRequest = function (sessionId) {
  return {
    'sessionId': sessionId,
    'action': 'getResult'
  }
};

/**
 * get submit result request
 * @param {string} sessionId
 * @return {{sessionId: string, action: string}}
 */
Hangman.prototype.generateSubmitResultRequest = function (sessionId) {
  return {
    'sessionId': sessionId,
    'action': 'submitResult'
  }
};

/**
 * load all dictionary
 */
Hangman.prototype.loadAllDict = function () {
  var self = this;
  this.option_.dictList.forEach(function (name) {
    self.loadDict(name);
  })
};

/**
 * load dictionary to memory
 *
 * @param {string} name
 */
Hangman.prototype.loadDict = function (name) {
  var self = this;
  var dictArray = fs.readFileSync(path.join(self.option_.dictPath, name), {encoding: 'utf8'}).split('\n'); //todo
  var index = self.dict_.length;
  self.dict_[index] = [];
  dictArray.forEach(function (word) {
    if (!self.dict_[index][word.length]) {
      self.dict_[index][word.length] = [];
    }
    self.dict_[index][word.length].push(word);
  });
  console.log('dictionary loaded: ' + name);
};

/**
 * update dictionary to higher level
 */
Hangman.prototype.upgradeSubDict = function () {
  this.currentDictLevel_ += 1;
  console.log('upgrade dictionary level: ' + this.currentDictLevel_);
  this.buildSubDictOfLength(this.gameContext_['currentWord'].length);
  this.reduceCurrentSubDictByWord(this.gameContext_['currentWord']);
  this.reduceCurrentSubDictByChars(this.guessedChar_);
};

/**
 * check whether can update dictionary
 * @return {boolean}
 */
Hangman.prototype.canUpgradeDict = function () {
  return this.dict_.length > this.currentDictLevel_ + 1;
};

/**
 * build sub-dictionary of specific length
 * @param {number} length
 */
Hangman.prototype.buildSubDictOfLength = function (length) {
  if (!this.dict_[this.currentDictLevel_]) {
    this.dict_[this.currentDictLevel_][length] = [];
  }
  this.currentSubDict_ = this.dict_[this.currentDictLevel_][length].slice();
};

/**
 * reduce the sub-dictionary by remove all word that don't match current word's pattern
 * @param {string} toBeGuessed
 */
Hangman.prototype.reduceCurrentSubDictByWord = function (toBeGuessed) {
  var negatedSetReg = '[^';
  for (var i = 0; i < toBeGuessed.length; ++i) {
    if (negatedSetReg.indexOf(toBeGuessed[i]) === -1 && toBeGuessed[i] !== '*') {
      negatedSetReg += toBeGuessed[i];
    }
  }
  negatedSetReg += ']';
  var regex = toBeGuessed.replace(/\*/g, negatedSetReg);
  this.currentSubDict_ = this.currentSubDict_.filter(function (word) {
    return word.match(regex);
  });
  if (this.currentSubDict_.length === 0) {
    console.warn('no candidate in dict: ' + this.gameContext_['currentWord']);
    if (this.canUpgradeDict()) {
      this.upgradeSubDict();
    } else {
      this.passCurrent_ = true;
    }
  }
};

/**
 * reduce the sub-dictionary by remove all word that has specific characters
 * @param {Array.<string>} characters
 */
Hangman.prototype.reduceCurrentSubDictByChars = function (characters) {
  this.currentSubDict_ = this.currentSubDict_.filter(function (word) {
    for (var i = 0; i < characters.length; ++i) {
      if (word.indexOf(characters[i]) === -1) {
        continue
      } else {
        return false;
      }
    }
    return true;
  });
  if (this.currentSubDict_.length === 0) {
    console.warn('no candidate in dict: ' + this.gameContext_['currentWord']);
    if (this.canUpgradeDict()) {
      this.upgradeSubDict();
    } else {
      this.passCurrent_ = true;
    }
  }
};

/**
 * get most possible guess for the word
 * @param {string} tobeGuessed
 * @return {string}
 */
Hangman.prototype.getMostPossibleChar = function (tobeGuessed) {
  var alphaMap = {'A': 0};
  var tmp = {};
  var result = 'A';

  this.currentSubDict_.forEach(function (word) {
    tmp = {};
    for (var i = 0; i < word.length; ++i) {
      if (tobeGuessed.indexOf(word[i]) === -1) {
        tmp[word[i]] = true;
      }
    }
    for (var alpha in tmp) {
      if (tmp.hasOwnProperty(alpha)) {
        alphaMap[alpha] = alphaMap[alpha] ? alphaMap[alpha] + 1 : 1;
        if (alphaMap[alpha] > alphaMap[result]) {
          result = alpha;
        }
      }
    }
  });
  return result;
};

/**
 * get next guess
 * @return {string}
 */
Hangman.prototype.guess = function () {
  var curWord = this.gameContext_['currentWord'];
  var guess = this.getMostPossibleChar(curWord);
  return guess;
};

/**
 * check whether guess on current word is finished
 * @return {boolean}
 */
Hangman.prototype.guessFinish = function () {
  return this.gameContext_['currentWord'].indexOf('*') === -1
};

/**
 * check whether last guess is correct
 * @return {boolean}
 */
Hangman.prototype.guessWrong = function () {
  return this.gameContext_['currentWord'] === this.gameContext_['previousWord'];
};

/**
 * check whether game is finished
 * @return {boolean}
 */
Hangman.prototype.allFinished = function () {
  return this.gameContext_['totalWordCount'] === this.gameContext_['numberOfWordsToGuess'];
};

/**
 * check whether guess to current word is failed
 * @return {boolean}
 */
Hangman.prototype.failed = function () {
  return this.gameContext_['wrongGuessCountOfCurrentWord'] === this.gameContext_['numberOfGuessAllowedForEachWord'];
};

